Patch to make Hugs output more suitable for use with hugs-hc and the
fptools testsuite:

* show less detail about pattern match failures to make output less brittle.

* print top-level exceptions in the same way as GHC-generated programs do.

--- Hugs/Prelude.hs	2004-11-17 10:39:58.000000000 +0000
+++ Hugs/Prelude.hs.new	2004-11-17 11:59:16.000000000 +0000
@@ -1633,7 +1633,7 @@
   showsPrec _ (IOException err)	  = shows err
   showsPrec _ (NoMethodError s)   = showException "undefined member" s
   showsPrec _ NonTermination	  = showString "<<loop>>"
-  showsPrec _ (PatternMatchFail s) = showException "pattern match failure" s
+  showsPrec _ (PatternMatchFail s) = showString "pattern match failure"
   showsPrec _ (RecConError s)     = showException "undefined field" s
   showsPrec _ (RecSelError s)     = showException "select of missing field" s
   showsPrec _ (RecUpdError s)     = showException "update of missing field" s
@@ -1929,11 +1929,15 @@
   exceptionHandler (ExitException ExitSuccess) = primExitWith 0
   exceptionHandler (ExitException (ExitFailure n)) = primExitWith n
   exceptionHandler err = runAndShowError $ do
-	putChar '\n'
-	putStr "Program error: "
-	putStrLn (show err)
+	hPutStr stderr "Fail: "
+	hPutStr stderr (show err)
+	hPutChar stderr '\n'
 	primExitWith 1
 
+primitive stderr      :: Handle
+primitive hPutChar      :: Handle -> Char -> IO ()
+primitive hPutStr       :: Handle -> String -> IO ()
+
 basicIORun :: IO a -> IOFinished a
 basicIORun (IO m) = loop [m hugsReturn]
 
